### PR TITLE
Template string improvements

### LINKED ISSUE #551: Template strings: Fixnums or floats with many decimals are rendered in exponential notation.
This PR https://github.com/garden-io/garden/pull/545 fixed an issue where module versions used in template strings were rendered as numbers in exponential format (instead of a standard string of decimals) if the first several characters in the commit hash were numbers.

To illustrate, here's an example from a failed CI run (from before https://github.com/garden-io/garden/pull/545 was merged):

screenshot 2019-02-18 at 14 06 22

Practically, this may only be a problem for versions / commit hashes, but we should keep an eye open for other bugs of this kind, in case there are other template-string related edge cases we haven't tackled yet.

### PR DESCRIPTION
**Pull request by edvald**
Look at individual commits for details. This addresses some issues with expected behavior, improves documentation a bit and paves the way for some features that would rely on templating e.g. booleans and numbers into configs.

Closes #551

### CODE REVIEW

File: garden-service/test/unit/src/template-string.ts [hunk 1]
```diff
@@ -209,12 +263,20 @@ describe("resolveTemplateString", async () => {
     expect(res).to.equal("foo")
   })
 
+  it("should handle chained conditional with string fallback", async () => {
```
Comment (thsig) [inline garden-service/test/unit/src/template-string.ts]: "handle chained" -> "handle a chained"?

File: garden-service/src/template-string.ts [hunk 2]
```diff
@@ -48,11 +52,21 @@ export async function resolveTemplateString(string: string, context: ConfigConte
...
   })
 
-  const resolved = await Bluebird.all(parsed)
-  return resolved.join("")
+  const resolved: (Primitive | undefined)[] = await Bluebird.all(parsed)
+
+  const result = resolved.length === 1
+    // Return value directly if there is only one value in the output
+    ? resolved[0]
+    // Else join together all the parts as a string. Output null as a literal string and not an empty string.
```
Comment (thsig) [inline garden-service/src/template-string.ts]: "join together all the parts" -> "join all the parts together"?

Comment (thsig) [inline garden-service/src/template-string.ts]: "and not an empty string" -> "and not as an empty string"?

File: garden-service/src/config/config-context.ts [hunk 3]
```diff
@@ -173,10 +174,26 @@ class LocalContext extends ConfigContext {
...
+  )
+  public username: () => Promise<string>
+
   constructor(root: ConfigContext) {
     super(root)
     this.env = process.env
     this.platform = process.platform
+    this.username = async () => {
+      const name = await username()
+      if (name === undefined) {
```
Comment (thsig) [inline garden-service/src/config/config-context.ts]: Should this be `if (!name) {` to catch `null` as well? Or is that unnecessary here?

Comment (edvald) [reply garden-service/src/config/config-context.ts]: That's unnecessary here. The function will either return a string or undefined.

File: garden-service/src/template-string.ts (hunk 2 above)
Comment (edvald) [reply garden-service/src/template-string.ts]: This is a code comment. No need to poke at the grammar .)

File: docs/using-garden/configuration-files.md [hunk 4]
```diff
@@ -313,10 +316,124 @@ tests:
...
+    namespace: ${local.env.CI_BRANCH || local.username || "default"}
+```
+
+This allows you to easily set default values when certain template keys are not available, and to configure your
+project based on varying context.
+
+### Numbers, booleans and null values
+
+When a template string key resolves to a number, boolean or null, the output is handled in two different ways,
+depending on whether the template string is part of surrounding string or not.
```
[the remark this answers is not in the document]
Comment (edvald) [reply docs/using-garden/configuration-files.md]: I don't think that's any clearer. But there is a missing article there, I'll fix that.

File: docs/using-garden/configuration-files.md [hunk 5]
```diff
@@ -313,10 +316,124 @@ tests:
...
+    log-level: ${local.env.LOG_LEVEL || "info"}
+    namespace: ${local.env.CI_BRANCH || local.username || "default"}
+```
+
+This allows you to easily set default values when certain template keys are not available, and to configure your
+project based on varying context.
+
+### Numbers, booleans and null values
+
+When a template string key resolves to a number, boolean or null, the output is handled in two different ways,
```
[the remark this answers is not in the document]
Comment (edvald) [reply docs/using-garden/configuration-files.md]: Why?

File: garden-service/src/template-string.ts (hunk 2 above)
Comment (thsig) [reply garden-service/src/template-string.ts]: Haha, no indeed—got a bit carried away there.

File: docs/reference/template-strings.md [hunk 6]
```diff
@@ -1,53 +1,53 @@
...
-services:
-  - name: some-service
-    # ...
-    env:
-      OTHER_MODULE_VERSION: ${modules.other-module.version}
-```
-
-You can also inject a template variable into a string. For instance, you might need to include a module's
-version as part of a URI:
+The following keys are available in template strings anywhere in Project `garden.yml` config files:
```
[the remark this answers is not in the document]
Comment (edvald) [reply docs/reference/template-strings.md]: This is meant to distinguish this section from the next section below, I'll clarify the language a bit.

**Pull request merged**